sphinx/themes/basic/static/searchtools.js

```diff
@@ -293,10 +298,12 @@ const Search = {
       if (title.toLowerCase().trim().includes(queryLower) && (queryLower.length >= title.length/2)) {
         for (const [file, id] of foundTitles) {
           let score = Math.round(100 * queryLower.length / title.length)
+          let isDocumentTitle = titles[file] === title
+          let anchor = id ? `#${id}` : ""
           results.push([
             docNames[file],
-            titles[file] !== title ? `${titles[file]} > ${title}` : title,
-            id !== null ? "#" + id : "",
+            isDocumentTitle ? title : `${titles[file]} > ${title}`,
+            isDocumentTitle ? "" : anchor, // don't provide an anchor if we matched on the exact document title (creates duplicates below)
```

Comment:
I think this makes sense, just modified my approach to not de-duplicate on anchor, just on filename, title, and description (the latter is just for object definitions). This will consider two entries in the same document with the same section title but different anchors duplicates, but I think that's probably acceptable. If that's the case I don't think the result is going to be particularly meaningful.

The alternative would be to pass down whether a result refers to the document title or not and then ignore the anchor if so, but that seemed significantly more complicated for little benefit.

Reply:
Interesting case where we have titles but unique sections:

https://docs.sqlalchemy.org/en/20/search.html?q=sql&check_keywords=yes&area=default

The top results are:

https://docs.sqlalchemy.org/en/20/changelog/changelog_03.html#change-0.3.0-sql
https://docs.sqlalchemy.org/en/20/changelog/changelog_03.html#change-0.3.4-sql
https://docs.sqlalchemy.org/en/20/changelog/changelog_03.html#change-0.3.5-sql

All of them have the same title `0.3 Changelog > sql`. With this change, we'd only select the top result. I think that would probably actually be kind of desirable? Prevents the ridiculous proliferation of nearly-identical search results (6551 on that page as of this writing)

Reply:
I changed my mind on this after thinking about it a bit more. I don't think it's a good idea to pull out valid unique search results out, so I went back and decided to do things the hard way -- I added an "isDocumentTitle" property to the search results, and omitted the anchor id from the de-duplication key if that's the case.

The end result passes the test, though it is a little awkward. Maybe there's a cleaner way of expressing this.

Reply:
I've opened https://github.com/wlach/sphinx/pull/1 with approximately what I had in mind for this - it does handle the `sqlalchemy` case you mention, although to be honest those search results would be clearer if we included the subsection headings (`0.3 > 0.3.11 > sql` for example).

Basically what it does is drop the `title` from the result de-duplication key.

Reply:
> I've opened https://github.com/wlach/sphinx/pull/1 with approximately what I had in mind for this - it does handle the sqlalchemy case you mention, although to be honest those search results would be clearer if we included the subsection headings (0.3 > 0.3.11 > sql for example).

Hmm, not sure this will work -- see my comment there. I might be missing something, but I don't think there's any way around special casing the "title" result.

> it does handle the sqlalchemy case you mention, although to be honest those search results would be clearer if we included the subsection headings (0.3 > 0.3.11 > sql for example).

Might be worth filing another issue for this. :smile: